internal/datacoord/garbage_collector.go

```diff
@@ -158,7 +163,35 @@ func (gc *garbageCollector) scan() {
 }
 
 func (gc *garbageCollector) clearEtcd() {
+	failedSegmentIDs := make([]int64, 0)
+	if gc.rcc == nil {
+		log.Warn("root coord client nil when garbage collecting")
+	} else {
+		req := &internalpb.GetImportFailedSegmentIDsRequest{
+			Base: &commonpb.MsgBase{
+				MsgType: commonpb.MsgType_GetImportFailedSegmentIDs,
+			},
+		}
+		ctx := context.Background()
+		rsp, err := gc.rcc.GetImportFailedSegmentIDs(ctx, req)
+		if err != nil {
+			log.Error("getImportFailedSegmentIDs failed", zap.String("error", err.Error()))
+		}
+		log.Debug("getImportFailedSegmentIDs from rootCoord succeed", zap.Int("IDNums", len(rsp.GetSegmentIDs())))
+		failedSegmentIDs = rsp.GetSegmentIDs()
+	}
+
+	failedSegmentIDsSet := make(map[int64]struct{})
+	for _, failedSegmentID := range failedSegmentIDs {
+		failedSegmentIDsSet[failedSegmentID] = struct{}{}
+	}
+
 	drops := gc.meta.SelectSegments(func(segment *SegmentInfo) bool {
+		_, failedSegmentFlag := failedSegmentIDsSet[segment.ID]
+		if failedSegmentFlag {
+			segment.State = commonpb.SegmentState_Dropped
```

Comment:
One more thing, do not update segment fields directly. Use copy on write:

https://github.com/milvus-io/milvus/blob/899a9bf86d16bc45422858fa2fff193f043b7c89/internal/datacoord/segment_info.go#L103-L107

Reply:
done